### PR TITLE
backport: Deprecate accounts

### PR DESCRIPTION
**Pull request by xdustinface**

### CODE REVIEW

**Comment by UdjinM6** [comment]
Would probably make sense to backport it together with https://github.com/bitcoin/bitcoin/pull/12892 and https://github.com/bitcoin/bitcoin/pull/13060.

File: src/rpc/client.cpp [hunk 1]
```diff
@@ -46,6 +46,7 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "getreceivedbyaddress", 2, "addlocked" },
     { "getreceivedbyaccount", 1, "minconf" },
     { "getreceivedbyaccount", 2, "addlocked" },
+    { "getreceivedbylabel", 2, "addlocked" },
```
Comment (UdjinM6) [inline src/rpc/client.cpp:49]:
```suggestion
    { "getreceivedbylabel", 1, "minconf" },
    { "getreceivedbylabel", 2, "addlocked" },
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/rpc/client.cpp (hunk 1 above)
Comment (xdustinface) [reply src/rpc/client.cpp:50]: Fixed and squashed

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -850,7 +885,9 @@ UniValue getbalance(const JSONRPCRequest& request)
             "Note that the account \"\" is not the same as leaving the parameter out.\n"
             "The server total may be different to the balance in the default \"\" account.\n"
             "\nArguments:\n"
-            "1. \"account\"        (string, optional) DEPRECATED. The selected account, or \"*\" for entire wallet. It may be the default account using \"\".\n"
+            "1. \"account\"        (string, optional) DEPRECATED. This argument will be removed in V0.18. \n"
+            "                     To use this deprecated argument, start dashd with -deprecatedrpc=accounts.\n"
+            "                     The selected account, or \"*\" for entire wallet. It may be the default account using \"\".\n"
```
Comment (UdjinM6) [inline src/wallet/rpcwallet.cpp:883]: It looks like all the params should be marked as deprecated, see https://github.com/bitcoin/bitcoin/pull/12953/files#diff-63c7710d72381cffe692175dd70e47f1891bf75c3e09249ed826f9df3917d02dR870

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -1060,9 +1108,48 @@ UniValue sendmany(const JSONRPCRequest& request)
         return NullUniValue;
     }
 
-    if (request.fHelp || request.params.size() < 2 || request.params.size() > 10)
-        throw std::runtime_error(
-            "sendmany \"fromaccount\" {\"address\":amount,...} ( minconf addlocked \"comment\" [\"address\",...] subtractfeefrom use_is use_ps conf_target \"estimate_mode\")\n"
+    std::string help_text;
+    if (!IsDeprecatedRPCEnabled("accounts")) {
+        help_text = "sendmany \"\" {\"address\":amount,...} ( minconf addlocked \"comment\" [\"address\",...] subtractfeefrom use_is use_ps conf_target \"estimate_mode\")\n"
+            "\nSend multiple times. Amounts are double-precision floating point numbers."
```
Comment (UdjinM6) [inline src/wallet/rpcwallet.cpp:1107]: missing https://github.com/bitcoin/bitcoin/pull/12953/files#diff-63c7710d72381cffe692175dd70e47f1891bf75c3e09249ed826f9df3917d02dR1093-R1094

File: test/functional/wallet_import_rescan.py [hunk 4]
```diff
@@ -120,23 +111,22 @@ def set_test_params(self):
...
     def setup_network(self):
-        extra_args = [['-deprecatedrpc=accounts'] for _ in range(self.num_nodes)]
+        extra_args = [[] for _ in range(self.num_nodes)]
         for i, import_node in enumerate(IMPORT_NODES, 2):
             if import_node.prune:
                 # txindex is enabled by default in Dash and needs to be disabled for import-rescan.py
                 extra_args[i] += ["-prune=1", "-txindex=0", "-reindex"]
 
-        self.add_nodes(self.num_nodes, extra_args=extra_args, stderr=sys.stdout)
+        self.add_nodes(self.num_nodes, stderr=sys.stdout)
```
Comment (UdjinM6) [inline test/functional/wallet_import_rescan.py:120]:
```suggestion
        self.add_nodes(self.num_nodes, extra_args=extra_args, stderr=sys.stdout)
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
`test/functional/wallet_listtransactions.py` is empty (`git mv` went wrong it seems) + see below

**Comment by xdustinface** [comment]
@UdjinM6 Good findings 👍 _Should_ have fixed/squashed them all now and `wallet_listtransactions.py` content is back and adjusted with the changes of bitcoin#13075.

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
Recent changes looks good 👍 however it looks like we need more backports to fix some issues introduced in already backported ones, at least https://github.com/bitcoin/bitcoin/pull/13498, https://github.com/bitcoin/bitcoin/pull/13566, https://github.com/bitcoin/bitcoin/pull/13130 and https://github.com/bitcoin/bitcoin/pull/14441.

**Comment by xdustinface** [comment]
@UdjinM6 Added the ones you mentioned. Will squash 692830b and 74a3513 into 9a3d528 before merge if someone prefers it, just left it out there yet for reviewing.

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -825,8 +868,9 @@ UniValue getbalance(const JSONRPCRequest& request)
         return NullUniValue;
     }
 
-    if (request.fHelp || request.params.size() > 4)
+    if (request.fHelp || (request.params.size() > 4 ))
```
Comment (UdjinM6) [inline src/wallet/rpcwallet.cpp:871]:
```suggestion
    if (request.fHelp || (request.params.size() > 4))
```

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -3781,20 +4116,16 @@ static const CRPCCommand commands[] =
...
     { "wallet",             "backupwallet",                     &backupwallet,                  {"destination"} },
     { "wallet",             "createwallet",                     &createwallet,                  {"wallet_name"} },
     { "wallet",             "dumpprivkey",                      &dumpprivkey,                   {"address"}  },
     { "wallet",             "dumpwallet",                       &dumpwallet,                    {"filename"} },
     { "wallet",             "encryptwallet",                    &encryptwallet,                 {"passphrase"} },
-    { "wallet",             "getaccountaddress",                &getaccountaddress,             {"account"} },
-    { "wallet",             "getaccount",                       &getaccount,                    {"address"} },
-    { "wallet",             "getaddressesbyaccount",            &getaddressesbyaccount,         {"account"} },
     { "wallet",             "getaddressinfo",                   &getaddressinfo,                {"address"} },
     { "wallet",             "getbalance",                       &getbalance,                    {"account","minconf","addlocked","include_watchonly"} },
```
Comment (UdjinM6) [inline src/wallet/rpcwallet.cpp:4126]:
```suggestion
    { "wallet",             "getbalance",                       &getbalance,                    {"account|dummy","minconf","addlocked","include_watchonly"} },
```
https://github.com/bitcoin/bitcoin/pull/13566/files#diff-63c7710d72381cffe692175dd70e47f1891bf75c3e09249ed826f9df3917d02dR4426

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
Looks good 👍 A couple of minor issues, see below.

File: src/rpc/client.cpp [hunk 7]
```diff
@@ -56,6 +58,9 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "listreceivedbyaccount", 1, "addlocked" },
     { "listreceivedbyaccount", 2, "include_empty" },
     { "listreceivedbyaccount", 3, "include_watchonly" },
+    { "listreceivedbylabel", 1, "addlocked" },
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply src/rpc/client.cpp:62]: Squashed into 1914effbbb5316ecf933fd24809eae0e67f3d419

File: test/functional/wallet_txn_doublespend.py [hunk 8]
```diff
@@ -62,9 +61,9 @@ def run_test(self):
         doublespend = self.nodes[0].signrawtransactionwithwallet(rawtx)
         assert_equal(doublespend["complete"], True)
 
-        # Create two spends using 1 500 DASH coin each
-        txid1 = self.nodes[0].sendfrom("foo", node1_address, 400, 0)
-        txid2 = self.nodes[0].sendfrom("bar", node1_address, 200, 0)
+        # Create two spends using 1 50 BTC coin each
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply test/functional/wallet_txn_doublespend.py]: Squashed into 42a98b0bc81dea90280c1dc07780bedd7fc18157

File: src/wallet/rpcwallet.cpp (hunk 6 above)
Comment (xdustinface) [reply src/wallet/rpcwallet.cpp]: Squashed into 341b5a9714200b69cef3d66056605bcefef93e54

File: src/wallet/rpcwallet.cpp (hunk 5 above)
Comment (xdustinface) [reply src/wallet/rpcwallet.cpp]: Squashed into 341b5a9714200b69cef3d66056605bcefef93e54

**Comment by xdustinface** [comment]
> A couple of minor issues, see below.

Squashed them all into the related commits 👍

**Review by UdjinM6 — APPROVED** [review]
LGTM

utACK

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
Please backport #14411 instead of #14441

Instead of backporting a backport, you should always backport the original PR instead, this makes life easier in the future.

(Probably fine to just change name it doesn't look like there's a diff)

**Comment by UdjinM6** [comment]
> Please backport bitcoin#14411 instead of bitcoin#14441

Unlike other backports, 14441 is more like a standalone fix because the changes to `src/wallet/rpcwallet.cpp` are significantly different from 14411 - it basically backports the idea, not the code. Backporting 14411 would require backporting many more PRs from 0.18 to merge cleanly it seems and it's probably why it was done this way in btc. I would propose to keep it as is and instead revert 96d3063 once we get to related 0.18 PRs later.

**Comment by PastaPastaPasta** [comment]
Ahh, gotcha, didn't realize that :+1:

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Pull request merged**